### ISSUE TITLE
XmlFrameDecoder is corrupt

### ISSUE THREAD

**Issue by tomaszc** [netty/netty#4030]
Netty Version: 4.1.0.Beta4

Context:
When writing TCP client which exchanges XML messages with server I've encountered an error in XmlFrameDecoder. Given not full XML element, decoder will output not valid XML.

Steps to reproduce:
1. BUG - Decode invalid XML: `'<a><b/></'` - decoder will put it to output list and it will update readerIndex in input byteBuff
2. OK - Decode invalid XML: `'<a><b/>'` - decoder will not put it to output list nor it will update readerIndex in input byteBuff
3. OK - Decode valid XML: `'<a><b/></a>'` - decoder will put it to output list and it will update readerIndex in input byteBuff

Long story short:
XML decoder assumes that as soon as it finds `</` it can decrement opened brackets count. If closing bracket isn't in byteBuff this assumption will create a bug.

**Comment by normanmaurer** [comment]
Can you also test with 4.1.0.Beta6-SNAPSHOT just to be sure it is not fixed already ?

**Comment by tomaszc** [comment]
I've just checked newest commit on branch 4.1 (94f65ed7ff214de3fc9680d80a5ec2d2fe2aaf48) and the problem still exists. 

The problem is in line 115, below comment *'found </, decrementing openBracketsCount'*. IMHO additional validation should be added there to check if closing bracket '>' is available in ByteBuf. I'll have to fix it, my application is broken beacouse of this bug - I'll create pull request ASAP.

**Comment by tomaszc** [comment]
@normanmaurer BTW - last version of netty on Maven central is 4.1.0.Beta4, isn't this some oversight?

**Comment by normanmaurer** [comment]
@tomaszc I can see Beta5 in maven central as well:
http://repo1.maven.org/maven2/io/netty/netty-all/4.1.0.Beta5/

**Comment by tomaszc** [comment]
@normanmaurer you're right, it is there, my mistake (morning without coffe ;) )

Meanwhile I've found additional error in XmlFrameDecoder:

StepsToReproduce:
Configure XmlFrameDecoder in your Channel Pipeline and send XML message with more than two root XML elements: `<a><b/><a><a><b/><a><a><b/><a>'. 
First decoder output will be proper: `<a><b/><a>`. 
Second decoder output will be corrupt: `<a><b/><a><a><b/><a>` instead of `<a><b/><a>`.

Cause:
In XmlFrameDecoder class, final part of decode(....) method is corrupt. When creating output frame (line 169) length parameter is invalid. It is based on input ByteBuff readerIndex, but it is not taken to account that it's value can be higher than 0. Therefore if XmlFrameDecoder will process byteBuff which readerIndex was already moved further, it's output will be corrupt.

**Comment by tomaszc** [comment]
Pull request with bugfix was created, hope it pass your verification
